core/apple/test/SourceNSInputStreamTest.kt

```diff
@@ -0,0 +1,354 @@
+                        } else {
+                            assertEquals("run-loop-2", NSThread.currentThread.name)
+                        }
+                        when (handleEvent) {
+                            NSStreamEventOpenCompleted -> opened.unlock()
+                            NSStreamEventHasBytesAvailable -> {
+                                if (read == 0) {
+                                    // switch to other run loop
+                                    input.removeFromRunLoop(runLoop1, NSDefaultRunLoopMode)
+                                    input.scheduleInRunLoop(runLoop2, NSDefaultRunLoopMode)
+                                } else if (read >= data.length - 3) {
+                                    // unsubscribe
+                                    input.removeFromRunLoop(runLoop2, NSDefaultRunLoopMode)
+                                }
+                                sink.usePinned {
+                                    val readBytes = input.read(it.addressOf(read).reinterpret(), 3U)
+                                    assertNotEquals(0, readBytes)
+                                    read += readBytes.toInt()
+                                }
+                                if (read == data.length) {
+                                    assertEquals(data, sink.decodeToString())
+                                    completed.unlock()
+                                }
+                            }
+                            NSStreamEventEndEncountered -> fail("$data shouldn't be subscribed")
+                        }
+                    }
+                }
+            }
+            input.scheduleInRunLoop(runLoop1, NSDefaultRunLoopMode)
+            input.open()
+            runBlocking {
+                opened.lockWithTimeout()
+                completed.lockWithTimeout()
+                // wait a bit to be sure delegate is no longer called
+                delay(200)
+            }
+            input.close()
+        }
+
+        consumeSwitching(NSInputStream(data = "default".encodeToByteArray().toNSData()), "default")
+        consumeSwitching(Buffer().apply { writeString("custom") }.asNSInputStream(), "custom")
+        CFRunLoopStop(runLoop1.getCFRunLoop())
+        CFRunLoopStop(runLoop2.getCFRunLoop())
+    }
+
+    @Test
+    fun testSubscribeAfterOpen() {
+        val runLoop = startRunLoop()
+
+        fun subscribeAfterOpen(input: NSInputStream) {
+            println("subscribeAfterOpen(${input::class})")
+            val available = Mutex(true)
+
+            input.delegate = object : NSObject(), NSStreamDelegateProtocol {
+                override fun stream(aStream: NSStream, handleEvent: NSStreamEvent) {
+                    println("${input::class} event = ${handleEvent.asString()} (${NSThread.currentThread.name})")
+                    assertEquals("run-loop", NSThread.currentThread.name)
+                    when (handleEvent) {
```

Comment:
Should the test fail on other events?

Reply:
Now after adding the data check after read, it will get the `NSStreamEventEndEncountered` event, which as long as it's in the correct order is expected. I'll add explicit checks for the other events to be thorough.